### ISSUE TITLE
After login, Nextcloud hangs, browsers report long-running script

### ISSUE THREAD

**Issue by Rhywden** [nextcloud/jsxc.nextcloud#119]
We currently have a huge user pool (1700+) with your chat module enabled. Upon login any user will encounter a freezing browser with the usual browser message that a long-running script is blocking everything. After stopping this script, Nextcloud will work as usual.

From profiling the browser performance after login, it seems to be the case that there are **two** forEach loops running and I'm guessing that it's thus looping twice (i. O(n²)) over all the users. Looks like this:

![grafik](https://user-images.githubusercontent.com/84586/51173974-f7156600-18b6-11e9-8b58-b18110cb1509.png)

You can easily see where the script is engaging / blocking @ 12800 ms.

Currently we don't have a stand-alone XMPP server behind our NextCloud instance.

Basically: What can we do to have a XMPP chat *without* this blocking issue?

**Comment by sualko** [comment]
Sorry for the late response. Your user base is really large, maybe too large for a contact list. Do you have any plans to reduce this number? I think it's not that helpful for your users, but we should fix this issue anyway. Can you try to reproduce this in our upcoming JSXC 4.0 release? https://www.jsxc.org/blog/2018/09/26/v4.0.0-first-alpha-released.html

**Comment by Rhywden** [comment]
Hey, thanks for the reply. I'm sorry to say that we will not be able to reduce the number of users as our Nextcloud instance is rather tightly coupled to our LDAP server and we do rather rely on the fliesharing capabilities as this enables easy access to files from outside the campus - something that's a strict requirement by our state for some educational curricula.
I'll talk to our sysadmin if he could give the alpha a whirl - I myself don't currently have admin access to that instance.
We're also thinking about setting up a dedicated XMPP server - would that also be able to mitigate the issue somewhat=

**Comment by sualko** [comment]
> we will not be able to reduce the number of users as our Nextcloud instance is rather tightly coupled to our LDAP server

I meant it's unlikely that everyone wants to chat with everyone on the campus. Therefore I suggest that you use an external XMPP server, so that every user can manage his contact list by it's own. Most XMPP servers also have an option or plugin to map ldap groups to the contact list. Hope this helps.

Nonetheless the insertion should be improved. Thanks for the hint.
